Fix: `gh pr create`, only fetch teams when reviewers contain a team

BagToad opened the pull request:
Fixes #11360 

See https://github.com/cli/cli/issues/11360#issuecomment-3103709170 for pre-implementation investigation and context.

This proposes a fix for the regression by reintroducing a condition to only fetch teams when a reviewer contains `/`. This proposes to maintain `gh pr edit`'s existing behavior by continuing to always fetch teams.

BagToad committed:
Add TeamReviewers flag to RepoMetadataInput

Introduces a TeamReviewers boolean to RepoMetadataInput to control whether team reviewers are fetched. Updates RepoMetadata logic to only fetch teams if both Reviewers and TeamReviewers are true. Adds tests to verify correct behavior when TeamReviewers is false.

BagToad committed:
FIX: conditionally fetching team reviewers

Updated the logic for fetching team reviewers in PR edit and create flows. In `pr edit`, team reviewers are always fetched for consistency with existing behavior, with a note to potentially align with `pr create` logic in the future. In `pr create`, team reviewers are now only fetched if a reviewer contains a slash, aligning with behavior before the regression.

BagToad commented on api/queries_repo_test.go:
```diff
@@ -213,6 +214,42 @@ func Test_RepoMetadata(t *testing.T) {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	require.NoError(t, err)
```
No I'll fix this a different way but good catch thanks

BagToad replied:
e5feda353f41b9be7b20af00bb80ba2032cc9bc7

BagToad committed:
Refactor error assertion in Test_RepoMetadataTeams

Replaces manual error check with require.NoError for consistency and improved readability in the test.

andyfeller commented on api/queries_repo.go:
```diff
@@ -964,7 +965,7 @@ func RepoMetadata(client *Client, repo ghrepo.Interface, input RepoMetadataInput
 	}
 
-	if input.Reviewers {
+	if input.Reviewers && input.TeamReviewers {
```
**question:**

1. what conditions will users still encounter the _`GraphQL: Resource not accessible by integration (organization.teams)`_ error?

   - GitHub CLI OAuth App?
   - GitHub Actions automatic token?

2. why is it important for both `Reviewers` and `TeamReviewers` to be set for retrieving organization teams?

BagToad replied:
1. I think it'll happen when someone gives us a team (contains a slash), but their auth token does not have perms to fetch teams.

2. It's technically not important to make this function but it eliminates a situation where the caller might set `TeamReviewers` to true while `Reviewers` is false, causing only teams to be fetched and populated into the reviewer metadata. I don't know of a case where we actually want that to happen, so I figure we should guard against callers misusing this field until we know that to be a valid use case by coupling the two fields together.

andyfeller replied:
Thanks for the insight!

For my first question, I was unsure which `gh` authentication methods include organization team permissions, so I reset my `gh auth` session and confirmed the  `read:org` scope associated with GitHub CLI OAuth app is good:

```shell
$ gh auth status --hostname github.com
github.com
  ✓ Logged in to github.com account andyfeller (keyring)
  - Active account: true
  - Git operations protocol: https
  - Token: gho_************************************
  - Token scopes: 'gist', 'read:org', 'repo', 'workflow'

$ QUERY='                             
query {
  organization(login:"cli"){
    teams(first:10, privacy:VISIBLE){
      nodes{
        slug
      }
    }
  }
}'

$ gh api graphql -f query="$QUERY"
{
  "data": {
    "organization": {
      "teams": {
        "nodes": [
          {
            "slug": "code-reviewers"
          },
          {
            "slug": "codespaces"
          },
          {
            "slug": "deployment-reviewers"
          },
          {
            "slug": "package-security"
          },
          {
            "slug": "tuf-root-reviewers"
          },
          {
            "slug": "webhooks"
          }
        ]
      }
    }
  }
}
```

andyfeller commented on pkg/cmd/pr/shared/params.go:
```diff
@@ -63,7 +64,10 @@ func AddMetadataToIssueParams(client *api.Client, baseRepo ghrepo.Interface, par
 	if tb.MetadataResult == nil {
 		input := api.RepoMetadataInput{
-			Reviewers:      len(tb.Reviewers) > 0,
+			Reviewers: len(tb.Reviewers) > 0,
+			TeamReviewers: len(tb.Reviewers) > 0 && slices.ContainsFunc(tb.Reviewers, func(r string) bool {
+				return strings.ContainsRune(r, '/')
+			}),
```
**question:** do we have any tests that should make sure that `OrganizationTeams` query is excluded when no teams are specified as reviewers?

I see tests like this that include teams:

https://github.com/cli/cli/blob/e5feda353f41b9be7b20af00bb80ba2032cc9bc7/pkg/cmd/pr/create/create_test.go#L412-L479

I only see 1 test that has only human users:

https://github.com/cli/cli/blob/e5feda353f41b9be7b20af00bb80ba2032cc9bc7/pkg/cmd/pr/create/create_test.go#L1300-L1382

If so, then maybe we `.Exclude(...)` that query from this case?

BagToad replied:
This test added in the PR removes the `OrganizationTeams` query mock while setting `TeamReviewers` to false. This would fail if the `OrganizationTeams` query actually runs because it won't find a mock for that query. 

https://github.com/cli/cli/blob/e5feda353f41b9be7b20af00bb80ba2032cc9bc7/api/queries_repo_test.go#L217-L247

andyfeller replied:
Thanks 🤔 I guess I was thinking we have a stronger assertion setup to ensure the query was explicitly not called at all close to the `gh pr create` command since it conditionally includes team reviewers whereas `gh pr edit` always does:

https://github.com/cli/cli/blob/b2348f838673786ab7d2a6ec577de81de3702b4b/api/queries_repo_test.go#L287-L300

BagToad replied:
Okay I spent some time firming up these tests, and I think they're well in the direction you mentioned. I ensured there are tests at the metadata fetcher level _and_ higher up in the `pr create` tests directly. This should cover all the bases.

addee16531fb834e7459bc9da105b0f551e5f71d
5a6cac3643df856d56d980b50cd3bb4352b64156

andyfeller reviewed (commented):
Great deep dive into the problem and fix! 🙌

Beyond the questions below, my last one is whether `gh pr create` and `gh pr edit` should have information about the permissions needed for using teams as reviewers 🤔

BagToad commented:
> Great deep dive into the problem and fix! 🙌
> 
> Beyond the questions below, my last one is whether `gh pr create` and `gh pr edit` should have information about the permissions needed for using teams as reviewers 🤔 

I think you're referring to the help docs for those two commands? I don't think it would hurt to document it.

andyfeller commented:
> > Beyond the questions below, my last one is whether `gh pr create` and `gh pr edit` should have information about the permissions needed for using teams as reviewers 🤔
> 
> I think you're referring to the help docs for those two commands? I don't think it would hurt to document it.

You got it. 🙇 Just an errant thought as I was reviewing the usage statement as I didn't see any mention of `read:org` scope for working with team reviewers.

https://github.com/cli/cli/blob/b2348f838673786ab7d2a6ec577de81de3702b4b/pkg/cmd/pr/create/create.go#L229-L233

https://github.com/cli/cli/blob/b2348f838673786ab7d2a6ec577de81de3702b4b/pkg/cmd/pr/edit/edit.go#L61-L62

andyfeller reviewed (approved):
Looks good to me 💪

BagToad committed:
Refactor and improve RepoMetadata teams test

Renamed the test to clarify its purpose and added an explicit exclusion for the OrganizationTeamList GraphQL query to ensure teams are only fetched when specified. This improves test accuracy and readability.

BagToad committed:
Add tests for reviewer team handling in PR creation

Added test cases to verify that teams are fetched when reviewers include teams and not fetched when only users are specified. This ensures correct behavior when requesting reviews from users and teams during pull request creation.

BagToad commented:
I wanted to also demo that this fix works on the originally reported GHES version `3.14.7`:

Reproduction with released binary:

<img width="1240" height="670" alt="image" src="https://github.com/user-attachments/assets/db0ad7e9-8053-4089-87e7-d810191d6755" />

Reproduction with binary built from this branch:

![fixed-ghes](https://github.com/user-attachments/assets/bf558c0a-2ed3-4ca8-8e6d-b58dafbebd1b)

andyfeller commented on pkg/cmd/pr/create/create_test.go:
```diff
@@ -1654,6 +1654,125 @@ func Test_createRun(t *testing.T) {
+				opts.BodyProvided = true
+				opts.Title = "my title"
+				opts.Body = "my body"
+				opts.Reviewers = []string{"hubot", "monalisa", "org/core", "org/robots"}
```
**praise:** I appreciate having the org added rather than leaving it clipped as in other tests

andyfeller commented on api/queries_repo_test.go:
```diff
@@ -242,6 +242,11 @@ func Test_RepoMetadataTeams(t *testing.T) {
 		  { "data": { "viewer": { "login": "monalisa" } } }
 		`))
 
+	http.Exclude(
+		t,
+		httpmock.GraphQL(`query OrganizationTeamList\b`),
+	)
+
```
**praise:** thank you for considering the idea ❤️

andyfeller reviewed (approved):
🦸 thank you for the due diligence!  ship it

jordi-275 reviewed (commented):
Abre

jordi-275 reviewed (commented):
Abre
